### PR TITLE
e2e: add initial support to multiple CNI

### PR DESCRIPTION
**Pull request by dougsland**
PLEASE NOTE: this branch is based in the PR: #169 (require merge first)

Testing KPNG with multiple CNI will increase the number of people
in the adoption road.
    
Signed-off-by: Douglas Schilling Landgraf <dlandgra@redhat.com>

### CODE REVIEW

File: hack/test_e2e.sh [hunk 1]
```diff
@@ -535,25 +536,41 @@ function help {
...
+        n ) cni_selected="${OPTARG}" ;;
         ? ) help ;; #Print help
     esac
 done
 
+if [[ "${cni_selected}" = false ]]; then
+    cni_selected="kindnet"
+fi
+
+if ! [[ "${cni_selected}" =~ ^(kindnet|calico|cilium|antrea|ovnkubernetes)$ ]]; then
```
Comment (knabben) [inline hack/test_e2e.sh:568]: to avoid confusion can we add only the ones we support and go incrementing according

File: hack/test_e2e.sh [hunk 2]
```diff
@@ -255,19 +255,15 @@ function create_cluster {
 
     echo -e "\nPreparing to setup ${E2E_CLUSTER_NAME} cluster ..."
     # create cluster
-    cat <<EOF | kind create cluster \
+
+    export E2E_IP_FAMILY
+    envsubst <"${0%/*}"/kindYAMLTemplate/"${cni_selected}"/kindConfig.yaml.tmpl > hack/kindYAMLTemplate/"${cni_selected}"/kindConfig.yaml
+    if_error_exit "cannot find kindconfig templte for CNI: ${cni_selected}"
```
Comment (knabben) [inline hack/test_e2e.sh:261]:
```suggestion
    if_error_exit "cannot find kindconfig template for CNI: ${cni_selected}"
```

**Comment by knabben** [comment]
@jayunit100 @dougsland 
What's the odd to integrate `https://github.com/thekubeworld/k8s-local-dev`, instead of rewriting it? I'm in favor to bring it upstream, not sure if this is the repo tough.

File: hack/test_e2e.sh (hunk 2 above)
Comment (dougsland) [reply hack/test_e2e.sh:261]: +1

File: hack/test_e2e.sh (hunk 1 above)
Comment (dougsland) [reply hack/test_e2e.sh]: sure

**Comment by dougsland** [comment]
> @jayunit100 @dougsland What's the odd to integrate `https://github.com/thekubeworld/k8s-local-dev`, instead of rewriting
> it? I'm in favor to bring it upstream, not sure if this is the repo tough.

You are right, I was about to re-write most of logic here. We can move k8s-local-dev to a different repo so more people own it too. It's up to you guys, looking forward to add multiple CNI support to test_e2e.

**Comment by jayunit100** [comment]
So I originally wrote kind-local-up.sh to have a modular CNI installer.

But, the broader KPNG group  made a decision, to remove that pluggable CNI installer, a while back, and antonio spent quite a bit of time removing our original calico/antrea based installer (kind-local-up.sh) so that it aligned with what we would need to run for an upstream prow job.
 
The reason was that, we have predicted 100+ issues that are going to be known bugs, and testing *n* CNIs simply isnt something that the repository needs to do to get us to the 
tactical deliverable of "replacing the kube-proxy".

Thus in the interest of really having that laser focus in place on the short/medium term deliverable, i'd say this is a lateral step and the better thing to focus on would be to find bugs (for example, with calico) and file them as one off issues, where they exist.

**Comment by dougsland** [comment]
> So I originally wrote kind-local-up.sh to have a modular CNI installer.
> 
> But, the broader KPNG group made a decision, to remove that pluggable CNI installer, a while back, and antonio spent quite a bit of time removing our original calico/antrea based installer (kind-local-up.sh) so that it aligned with what we would need to run for an upstream prow job.
> 
> The reason was that, we have predicted 100+ issues that are going to be known bugs, and testing _n_ CNIs simply isnt something that the repository needs to do to get us to the tactical deliverable of "replacing the kube-proxy".
> 
> Thus in the interest of really having that laser focus in place on the short/medium term deliverable, i'd say this is a lateral step and the better thing to focus on would be to find bugs (for example, with calico) and file them as one off issues, where they exist.


Make sense Jay, let's put this one on hold. I am trying to help in several fronts this one will be in hold.

**Comment by dougsland** [comment]
Opened an issue for someone that would like to continue the work. Closing for now.
https://github.com/kubernetes-sigs/kpng/issues/225

**Pull request closed without merging**